README.md
# WHSacc

<!-- badges: start -->

<!-- badges: end -->

The goal of WHSacc is to automatically and interactively process the [Women's Health Study (WHS)](https://whs.bwh.harvard.edu/) accelerometer data files using the [GGIR](https://cran.r-project.org/web/packages/GGIR/index.html) package.

## Installation and use example

You can install the development version of WHSacc from [GitHub](https://github.com/) with:

``` r
# install.packages("devtools")
devtools::install_github("jhmigueles/WHSacc")
```

To use the WHSacc package, run the "runGGIR" function and follow the instructions:

``` r
library(WHSacc)
WHSacc::runGGIR()
```

## Accelerometer data collection protocol

This is a brief summary on how the accelerometer data was collected in the WHS:

-   **Device:** ActiGraph GT3X
-   **Attachment site:** hip
-   **Sample frequency:** 30 Hz
-   **Protocol:** devices mailed to participants, who were asked to wear them for 7 consecutive days, removing them for sleep and water-based activities
-   **Files to process:** gt3x files zipped into either ".gz" or ".7z" files.

## How this package works

1.  The user is asked to select the files to process
2.  The files are unzipped and stored in a new directory named "unzipped files"
3.  GGIR is called and the data processing begins
    -   if there are more than 100 files, the files are processed in chunks of 100 with a for loop
    -   Output and datasets are derived at the end of the for loop, so, if the process is interrupted before ending the for loop, you will not see the GGIR datasets and visualizations
4.  If specified by the user, the generated files in the "unzipped files" directory are removed at the end of the process

## GGIR function call

Summary of the settings used:

-   Epoch length of 5 seconds and aggregated to 60 seconds in g.part5
-   At least 10 wear hours to consider a valid day
-   Sleep period time fixed from 11PM to 6AM
-   [Hildebrand et al. cutoffs for hip](https://pubmed.ncbi.nlm.nih.gov/24887173/)
-   Sedentary bouts of \>30 and \>60 min/day
-   MVPA bouts of \>5 and \>10 min/day

This is the function call I use within the WHSacc package:

``` r
GGIR(mode = 1:5,
     #BASIC SETTINGS
     datadir = datadir,         # selected by user
     outputdir = outputdir,     # selected by user
     studyname = "WHS",
     overwrite = overwrite,     # selected by user
     print.filename = T, printsummary = T,
     sensor.location = "hip",
     #G.PART1: GET ENMO VALUES
     windowsizes = c(5, 900, 3600),
     desiredtz = "US/Eastern",
     do.enmo = T, do.mad = F,
     do.anglex = T, do.angley = T, do.anglez = T,
     dayborder = 0,
     acc.metric = "ENMO",
     #G.PART2: FIRST ESTIMATIONS OF PHYSICAL ACTIVITY
     idloc = 6,
     strategy = 1,
     hrs.del.start = 0, hrs.del.end = 0,
     includedaycrit = 10,
     winhr = c(5,10),
     qwindow = c(0, 24),
     qlevels = c((1440 - 60)/1440,   # M60
                 (1440 - 30)/1440,   # M30
                 (1440 - 15)/1440,   # M15
                 (1440 - 10)/1440,   # M10
                 (1440 - 5)/1440),   # M5
     ilevels = c(),
     mvpathreshold = 70,
     boutcriter = 0.8, bout.metric = 6,
     epochvalues2csv = FALSE,
     #G.PART3-4: SLEEP CLASSIFICATION
     def.noc.sleep = c(23, 6),
     relyonguider = TRUE,
     sleepwindowType = "SPT",
     constrain2range = TRUE,
     HASPT.algo = c(),
     longitudinal_axis = 1,
     anglethreshold = 5, timethreshold = 5,
     possible_nap_window = c(9, 18),
     possible_nap_dur = c(30, 180),
     do.visual = TRUE,
     outliers.only = FALSE,
     includenightcrit = 0,
     ignorenonwear = FALSE,
     #G.PART5: MERGE SLEEP AND PHYSICAL ACTIVITY DATA
     part5_agg2_60seconds = TRUE,
     threshold.lig = 20, threshold.mod = 70, threshold.vig = 260, 
     boutdur.mvpa = c(5, 10), boutdur.in = c(30, 60), boutdur.lig = c(10, 30),
     boutcriter.mvpa = 0.8, boutcriter.in = 0.9, boutcriter.lig = 0.8,
     timewindow = c("MM"),
     save_ms5rawlevels = TRUE, save_ms5raw_without_invalid = FALSE,
     save_ms5raw_format = "csv",
     week_weekend_aggregate.part5 = FALSE,
     #REPORTS
     do.report = c(2,4,5),
     visualreport = TRUE,
     do.parallel = TRUE)
```


NEWS.md
# WHSacc

\name{NEWS}
\title{News for Package \pkg{WHSacc}}

\section{Changes in version 1.0.0 [Baltimore] (Github-only release date: 31-7-2022)}{
\itemize{
  \item First running version of the package
}
}


R/unzipWHSfiles.R
#' unzipWHSfiles
#'
#' @description Function to decompress the WHS files (only works with gz or 7z files)
#'
#' @param WHSfiles directory with the files to unzip
#'
#' @return datadir definition for GGIR
#' @export
#'
#' @importFrom archive archive_extract
#' @importFrom R.utils gunzip
#'
unzipWHSfiles = function(WHSfiles) {

  # create new directory to unzip files
  datadir = file.path(dirname(WHSfiles[1]), "unzipped files")
  if (!dir.exists(datadir)) dir.create(datadir)
  cat("Extracting zipped files...\n")
  pb = txtProgressBar(min = 0, max = length(WHSfiles), initial = 0, style = 3)

  # keep unzipped files for new datadir
  keep = c()

  for (i in 1:length(WHSfiles)) {
    # find out file extension
    format = tools::file_ext(x = WHSfiles[i])

    # extract files in case they are gz or 7z
    if (format == "gz") {
      tmp1 = unlist(strsplit(WHSfiles[i], "[.]cs"))
      tmp2 = unlist(strsplit(WHSfiles[i], "[.]gt"))

      if (length(tmp1) > 1) { # then a csv file
        FixPath4GGIR = !grepl(pattern = "csv.gz", x = WHSfiles[i])
        if (FixPath4GGIR) {
          remove = gsub("v", "", tmp1[2])
          remove = gsub(".gz", "", remove)
          newname = gsub(remove, "", WHSfiles[i])
          file.rename(WHSfiles[i], newname)
          keep = c(keep, newname)
          rm(newname); gc()
        } else {
          keep = c(keep, WHSfiles[i])
        }
      } else if (length(tmp2) > 1 ) {
        # then a gt3x file
        # check that file extension is fine
        newname = gsub(".gz", "", WHSfiles[i])
        if (tools::file_ext(newname) != "gt3x") {
          newname = paste0(newname, ".gt3x")
        }
        newfile = file.path(datadir, basename(newname))

        if (!file.exists(newfile)) {
          R.utils::gunzip(WHSfiles[i], remove = FALSE,
                          destname = newfile)
        }
      }
    } else if (format == "7z") {
      archive::archive_extract(WHSfiles[i],
                               dir = datadir)
      # newname = gsub(".7z", "", WHSfiles[i])
      # # check that file extension is fine
      # if (tools::file_ext(newname) != "gt3x") {
      #   newname2 = paste0(newname, ".gt3x")
      #   file.rename(newname, newname2)
      # }
    } else {
      stop("This package only works with gz or 7z files at the moment.")
    }
    setTxtProgressBar(pb,i)
  }

  # define new datadir
  files2process = c(dir(datadir, full.names = TRUE), keep)
  files2process = gsub("\\", "/", files2process, fixed = TRUE)

  cat("\nFiles unzipped\n\n")

  invisible(files2process)

}


R/runGGIR.R
#' runGGIR
#'
#' @description Main function to run GGIR over the WHS files
#'
#' @param interactive If TRUE (default), the directory selection is interactive
#'
#' @export
#'
#' @import GGIR
runGGIR = function(interactive = TRUE) {

  # before starting, make sure we work with GGIR v.2.7.2 or newer and archive v.1.1.5
  if (packageVersion("GGIR") < "2.7.3" & packageVersion("archive") < "1.1.5") {
    cat("Please, update the GGIR and the archive packages before running WHSacc:\n
         1 - restart the R session\n
         2 - devtools::install_github('wadpac/GGIR')
         3 - install.packages('archive')\n")
    stop()
  } else if (packageVersion("GGIR") < "2.7.3" | packageVersion("archive") < "1.1.5") {
    package = ifelse(packageVersion("GGIR") < "2.7.3", "GGIR", "archive")
    stop(cat("Please, update", package, "before running WHSacc:\n",
             "1 - restart the R session\n",
             ifelse(package == "GGIR", "2 - devtools::install_github('wadpac/GGIR')",
                    "2 - install.packages('archive')\n")))
  }

  # 1 - select files
  if (interactive) {
    zipfiles = invisible(choose.files(caption = "Select the accelerometer files to process",
                                           multi = TRUE))
    zipfiles = gsub('\\\\', '/', zipfiles)
    zipdir = dirname(zipfiles)
  } else {
    zipdir = readline(prompt = 'Write the path to the directory with your files: ')
    zipdir = gsub('\\"', '', zipdir)
    zipdir = gsub("\\'", "", zipdir)

    zipfiles = dir(zipdir, full.names = TRUE)
    if ("unzipped files" %in% basename(zipfiles)) zipfiles = zipfiles[-which(basename(zipfiles) == "unzipped files")]
  }
  nFiles = length(zipfiles)
  cat(paste0("\nThere are ", nFiles,
             " files to process.\n",ifelse(nFiles > 6, "Here the first 6 file names", "Here are the file names"),
             ", is this correct?\n"))
  print(head(basename(zipfiles)))

  cat("\n1: Yes\n2: No\n")

  moveon = readline(prompt = 'Answer: ')

  if (moveon == "2") stop("Run again the runGGIR function and select the correct files")

  # From here on, run in chunks in case there are more than 100 files
  if (nFiles > 100) {
    f0 = seq(1, nFiles, by = 100)
    f1 = seq(100, nFiles, by = 100)
    if (length(f0) != length(f1)) f1[length(f1) + 1] = nFiles
  } else {
    f0 = 1
    f1 = nFiles
  }

  # This should never occur, but just in case:
  if (length(f0) != length(f1)) stop("Contact Jairo, something went wrong here")

  # 3 - outputdir
  cat("\n")
  cat(paste0(rep("_", options()$width), collapse = ""))
  outputdir = unique(dirname(zipdir))

  cat("\nShould the output files and datasets be stored in", paste0(outputdir,"?"))
  cat("\n1: Yes\n2: No\n")
  moveon = readline(prompt = 'Answer: ')
  if (moveon == "2") {
    outputdir = readline(prompt = 'Write the path to the directory to save the output files and datasets: ')
    outputdir = gsub('\\"', '', outputdir)
    outputdir = gsub("\\'", "", outputdir)
  }

  # 4 - should the files be removed at the end of the process?
  cat("\n")
  cat(paste0(rep("_", options()$width), collapse = ""))
  cat("\nDo you want to remove the unzipped files generated at the end of the process?")
  cat("\n1: Yes\n2: No\n")
  removeFiles = readline(prompt = 'Answer: ')
  cat("\n")

  for (chunk in 1:length(f0)) {
    # backup of all selected files
    if (chunk == 1) zipfiles_all = zipfiles

    # files to process in this loop
    zipfiles = zipfiles_all[f0[chunk]:f1[chunk]]

    # 1 - unzip files
    datadir = unzipWHSfiles(zipfiles)

    # 2 - run GGIR:
    # set overwrite to FALSE in case this is an interation of the for loop
    if (chunk > 1) overwrite = FALSE

    GGIR::GGIR(mode = 1:5,

               #BASIC SETTINGS
               datadir = datadir,
               outputdir = outputdir,
               studyname = "WHS",
               overwrite = overwrite,
               print.filename = T, printsummary = T,
               sensor.location = "hip",

               #G.PART1: GET ENMO VALUES
               windowsizes = c(5, 900, 3600),
               desiredtz = "US/Eastern",
               do.enmo = T, do.mad = F,
               do.anglex = T, do.angley = T, do.anglez = T,
               dayborder = 0,
               acc.metric = "ENMO",

               #G.PART2: FIRST ESTIMATIONS OF PHYSICAL ACTIVITY
               idloc = 6,
               strategy = 1,
               hrs.del.start = 0, hrs.del.end = 0,
               includedaycrit = 10,
               winhr = c(5,10),
               qwindow = c(0, 24),
               qlevels = c((1440 - 60)/1440,   # M60
                           (1440 - 30)/1440,   # M30
                           (1440 - 15)/1440,   # M15
                           (1440 - 10)/1440,   # M10
                           (1440 - 5)/1440),   # M5
               ilevels = c(),
               mvpathreshold = 70,
               boutcriter = 0.8, bout.metric = 6,
               epochvalues2csv = FALSE,

               #G.PART3-4: SLEEP CLASSIFICATION
               def.noc.sleep = c(23, 6),
               relyonguider = TRUE,
               sleepwindowType = "SPT",
               constrain2range = TRUE,
               HASPT.algo = c(),
               longitudinal_axis = 1,
               anglethreshold = 5, timethreshold = 5,
               possible_nap_window = c(9, 18),
               possible_nap_dur = c(30, 180),
               do.visual = TRUE,
               outliers.only = FALSE,
               includenightcrit = 0,
               ignorenonwear = FALSE,

               #G.PART5: MERGE SLEEP AND PHYSICAL ACTIVITY DATA (DEFINITIVE DATASET)
               part5_agg2_60seconds = TRUE,
               threshold.lig = 20, threshold.mod = 70, threshold.vig = 260,
               boutdur.mvpa = c(5, 10), boutdur.in = c(30, 60), boutdur.lig = c(10, 30),
               boutcriter.mvpa = 0.8, boutcriter.in = 0.9, boutcriter.lig = 0.8,
               timewindow = c("MM"),
               save_ms5rawlevels = TRUE, save_ms5raw_without_invalid = FALSE,
               save_ms5raw_format = "csv",
               week_weekend_aggregate.part5 = FALSE,

               #REPORTS
               do.report = c(2,4,5),
               visualreport = TRUE,
               do.parallel = TRUE)


    # 6 - remove processed files?

    if(removeFiles == "1") {
      cat("\n\nRemoving generated files...")
      remove = unique(dirname(datadir))
      remove = remove[which(basename(remove) == "unzipped files")]
      unlink(remove, recursive = TRUE)
    }

    # END!!
    cat("\n\nDone!\n\n")
  }
}
